plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/logging/QueryCommentContextAppender.java

```diff
@@ -0,0 +1,126 @@
+import static java.util.stream.Collectors.joining;
+
+public class QueryCommentContextAppender
+        implements QueryExecutionContextAppender
+{
+    private static final List<String> PREDEFINED_MATCHES = Arrays.stream(PredefinedValue.values()).map(PredefinedValue::getMatchCase).toList();
+    private static final List<String> PREDEFINED_VALUE_CODES = Arrays.stream(PredefinedValue.values()).map(PredefinedValue::getPredefinedValueCode).toList();
+    private static final Pattern VALIDATION_PATTERN = Pattern.compile("[\\w ,=]|" + String.join("|", PREDEFINED_MATCHES));
+    private final String commentFormat;
+
+    @Inject
+    public QueryCommentContextAppender(QueryCommentContextAppenderConfig config)
+    {
+        this(config.getFormat());
+    }
+
+    public QueryCommentContextAppender(String commentFormat)
+    {
+        this.commentFormat = requireNonNull(commentFormat, "commentFormat is null");
+        checkArgument(!commentFormat.isBlank(), "Empty format is not allowed");
+        Matcher matcher = VALIDATION_PATTERN.matcher(commentFormat);
+        checkArgument(
+                matcher.results()
+                        .map(MatchResult::group)
+                        .collect(joining())
+                        .equals(commentFormat),
+                "Incorrect format: %s. Format may consist of only letters, digits, underscores, commas, spaces, equal signs and predefined values: %s", commentFormat, String.join(", ", PREDEFINED_VALUE_CODES));
+    }
+
+    @Override
+    public String appendContext(String query, ConnectorSession session)
+    {
+        String message = commentFormat;
+        for (PredefinedValue predefinedValue : PredefinedValue.values()) {
+            message = message.replaceAll(predefinedValue.getMatchCase(), predefinedValue.value(session));
+        }
+        return new StringBuilder()
+                .append(query)
+                .append("/*")
+                .append(message)
+                .append("*/")
+                .toString();
+    }
+
+    private enum PredefinedValue
+    {
+        QUERY_ID(ConnectorSession::getQueryId),
+        SOURCE(new SanitizedValuesProvider(session -> session.getSource().orElse(""), "$SOURCE")),
+        USER(ConnectorSession::getUser),
+        TRACE_TOKEN(new SanitizedValuesProvider(session -> session.getTraceToken().orElse(""), "$TRACE_TOKEN"));
+
+        private final Function<ConnectorSession, String> valueProvider;
+
+        PredefinedValue(Function<ConnectorSession, String> valueProvider)
+        {
+            this.valueProvider = valueProvider;
+        }
+
+        String getMatchCase()
+        {
+            return "\\$" + this.name();
+        }
+
+        String getPredefinedValueCode()
+        {
+            return "$" + this.name();
+        }
+
+        String value(ConnectorSession session)
+        {
+            return valueProvider.apply(session);
+        }
+    }
+
+    private static class SanitizedValuesProvider
```

Comment:
No, because the concept is to sanitize input data. So `USER` and `QUERY_ID` are not sanitized, as they don't have to, where `TRACE_TOKEN` and `SOURCE` needs to be checked to prevent sql injection attacks.